ConsoleApp/Program.cs

```diff
@@ -0,0 +1,267 @@
+        public class LinkedList<T> : IReadOnlyList<T>, IWriteOnlyList<T>
+        {
+            private class ListItem
+            {
+                public T Value { get; set; }
+                public ListItem Next { get; set; }
+            }
+            private ListItem _head;
+            public int Count { get; private set; }
+            public void Push(T item)
+            {
+                Console.WriteLine("Push: {0}", item);
+                Insert(item, Count);
+            }
+            public T this[int index]
+            {
+                get => GetByIndex(index).Value;
+                set
+                {
+                    var item = GetByIndex(index);
+                    item.Value = value;
+                }
+            }
+            public T[] GetAll()
+            {
+                var array = new T[Count];
+                var item = _head;
+
+                for (int i = 0; i < Count; ++i)
+                {
+                    array[i] = item.Value;
+                    item = item.Next;
+                }
+
+                return array;
+            }
+            public void Clear()
+            {
+                _head = null;
+                Count = 0;
+            }
+            private T ShowAndRemove(int index)
+            {
+                T retValue;
+                if ((index == Count) || (index < 0))
+                {
+                    // throw new ArgumentOutOfRangeException(nameof(index));
+                    throw new Exception("Stack is Empty");
+                }
+
+                if (index == 0)
+                {
+                    var prev = GetByIndex(0);
+                    retValue = prev.Value;
+                    _head = _head.Next;
+                }
+                else
+                {
+                    var prev = GetByIndex(index - 1);
+                    retValue = prev.Next.Value;
+                    prev.Next = prev.Next.Next;
+                }
+
+                --Count;
+                return retValue;
+            }
+            private T Show(int index)
+            {
+                T retValue;
+                if ((index == Count) || (index < 0))
+                {
+                    throw new Exception("Stack is Empty");
+                    // throw new ArgumentOutOfRangeException(nameof(index));
+                }
+
+                if (index == 0)
+                {
+                    var prev = GetByIndex(0);
+                    retValue = prev.Value;
+                }
+                else
+                {
+                    var prev = GetByIndex(index - 1);
+                    retValue = prev.Next.Value;
+                }
+                return retValue;
+            }
+            public T Pop()
+            {
+                try
+                {
+                    return ShowAndRemove(Count - 1);
+                }
+                catch (Exception ex)
+                {
+                    return default(T);
```

Comment:
if you have no elements it's better to throw this exception because your stack can have elements equal to the default